dali/kernels/imgproc/convolution/separable_convolution_cpu.h

```diff
@@ -0,0 +1,202 @@
+                "Only floats as intermediate values are currently supported.");
+};
+
+template <typename Out, typename In, typename W, bool has_channels>
+struct SeparableConvolutionCpuImpl<Out, In, W, 3, has_channels> {
+  static constexpr int axes = 3;
+  static constexpr int ndim = has_channels ? 4 : 3;
+
+  KernelRequirements Setup(KernelContext& ctx, const TensorShape<ndim>& in_shape,
+                           const std::array<int, axes>& window_sizes) {
+    KernelRequirements req;
+
+    ScratchpadEstimator se;
+    se.add<W>(AllocType::Host, volume(in_shape));
+    req.scratch_sizes = se.sizes;
+    req.output_shapes.push_back(uniform_list_shape<ndim>(1, in_shape));
+
+    auto req_inner = conv_innermost_.Setup(ctx, in_shape, window_sizes[2]);
+    auto req_middle = conv_middle_.Setup(ctx, in_shape, window_sizes[1]);
+    auto req_outer = conv_outermost_.Setup(ctx, in_shape, window_sizes[0]);
+
+    req.AddInputSet(req_inner, false);
+    req.AddInputSet(req_middle, false);
+    req.AddInputSet(req_outer, false);
+
+    return req;
+  }
+
+  void Run(KernelContext& ctx, const TensorView<StorageCPU, Out, ndim> out,
+           const TensorView<StorageCPU, const In, ndim>& in,
+           const std::array<TensorView<StorageCPU, const W, 1>, axes>& windows,
+           const std::array<W, axes>& scales = uniform_array<axes, W>(1.f)) {
+    auto* tmp = ctx.scratchpad->Allocate<W>(AllocType::Host, volume(in.shape));
+    auto intermediate = GetInterStageView(out, tmp);
+
+    conv_innermost_.Run(ctx, intermediate, in, windows[2], scales[2]);
+    conv_middle_.Run(ctx, intermediate, intermediate, windows[1], scales[1]);
+    conv_outermost_.Run(ctx, out, intermediate, windows[0], scales[0]);
+  }
+
+  ConvolutionCpu<W, In, W, ndim, 2, has_channels> conv_innermost_;
+  ConvolutionCpu<W, W, W, ndim, 1, has_channels> conv_middle_;
+  ConvolutionCpu<Out, W, W, ndim, 0, has_channels> conv_outermost_;
+};
+
+template <typename Out, typename In, typename W>
+struct SeparableConvolutionCpu<Out, In, W, 1, false>
+    : public SeparableConvolutionCpuImpl<Out, In, W, 1, false> {};
+
+template <typename Out, typename In, typename W>
+struct SeparableConvolutionCpu<Out, In, W, 2, true>
+    : public SeparableConvolutionCpuImpl<Out, In, W, 1, true> {};
+
+template <typename Out, typename In, typename W>
+struct SeparableConvolutionCpu<Out, In, W, 2, false>
+    : public SeparableConvolutionCpuImpl<Out, In, W, 2, false> {};
+
+template <typename Out, typename In, typename W>
+struct SeparableConvolutionCpu<Out, In, W, 3, true>
+    : public SeparableConvolutionCpuImpl<Out, In, W, 2, true> {};
+
+template <typename Out, typename In, typename W>
+struct SeparableConvolutionCpu<Out, In, W, 3, false>
+    : public SeparableConvolutionCpuImpl<Out, In, W, 3, false> {};
+
+template <typename Out, typename In, typename W>
+struct SeparableConvolutionCpu<Out, In, W, 4, true>
+    : public SeparableConvolutionCpuImpl<Out, In, W, 3, true> {};
```

Comment:
What is it for? Why not just rename XxxImpl to Xxx?

Reply:
So the Kernel is parametrized with the number of actual dimensions and not only data dimensions + bool for the channels.

Reply:
It's a personal opinion, but I find it _harder_ to use this way - I mean, I find it more intuitive to parameterize 2D convolution with channels and without channels with <2>, not <2> or <3>. From the operator standpoint, you still need to handle it explicitly, either in value switch or an `if`, so there's no difference.

Reply:
Removed, now it's less code which has some benefits.